File: nostr-knowledge/libraries/dart/examples/basic_usage.dart
/// Nostr Dart/Flutter Basic Usage Example
///
/// Demonstrates key generation, event creation, relay connection,
/// publishing, and subscribing using dart_nostr.
///
/// pubspec.yaml dependencies:
///   dart_nostr: ^9.2.4
///
/// For NDK and dart-nostr alternatives, see sections below.

import 'dart:async';
import 'package:dart_nostr/dart_nostr.dart';

// =============================================================================
// Section 1: Using dart_nostr (anasfik/nostr)
// =============================================================================

Future<void> main() async {
  print('=== Nostr Dart Example (dart_nostr) ===\n');

  final nostr = Nostr.instance;

  // ---------------------------------------------------------------------------
  // 1. Key Generation
  // ---------------------------------------------------------------------------

  // Generate a new random keypair
  final keyPair = nostr.keysService.generateKeyPair();
  print('Public key (hex):  ${keyPair.public}');
  print('Private key (hex): ${keyPair.private}');

  // Encode to bech32
  final npub = nostr.keysService.encodePublicKeyToNpub(keyPair.public);
  final nsec = nostr.keysService.encodePrivateKeyToNsec(keyPair.private);
  print('Public key (npub): $npub');
  print('Secret key (nsec): $nsec');

  // Derive public key from private key
  final derivedPubKey = nostr.keysService.derivePublicKey(
    privateKey: keyPair.private,
  );
  assert(derivedPubKey == keyPair.public);
  print('Key derivation verified.');
  print('');

  // ---------------------------------------------------------------------------
  // 2. Relay Connection
  // ---------------------------------------------------------------------------

  print('Connecting to relays...');
  await nostr.relaysService.init(
    relaysUrl: [
      'wss://relay.damus.io',
      'wss://nos.lol',
      'wss://relay.nostr.band',
    ],
  );
  print('Connected to ${nostr.relaysService.relaysList?.length ?? 0} relays.');
  print('');

  // ---------------------------------------------------------------------------
  // 3. Event Creation and Signing
  // ---------------------------------------------------------------------------

  // Create a text note (kind 1)
  final textNote = NostrEvent.fromPartialData(
    kind: 1,
    content: 'Hello Nostr from Dart!',
    keyPairs: keyPair,
    tags: [
      ['t', 'nostr'],
      ['t', 'dart'],
    ],
  );

  print('Event ID:        ${textNote.id}');
  print('Event kind:      ${textNote.kind}');
  print('Event content:   ${textNote.content}');
  print('Event signature: ${textNote.sig?.substring(0, 32)}...');
  print('');

  // Create a metadata event (kind 0)
  final metadataEvent = NostrEvent.fromPartialData(
    kind: 0,
    content: '{"name":"DartDev","about":"Building nostr clients in Dart","nip05":"dartdev@example.com"}',
    keyPairs: keyPair,
  );

  // Create a recommended relay list (kind 10002, NIP-65)
  final relayListEvent = NostrEvent.fromPartialData(
    kind: 10002,
    content: '',
    keyPairs: keyPair,
    tags: [
      ['r', 'wss://relay.damus.io', 'read'],
      ['r', 'wss://nos.lol', 'write'],
      ['r', 'wss://relay.nostr.band'],
    ],
  );

  // ---------------------------------------------------------------------------
  // 4. Publishing Events
  // ---------------------------------------------------------------------------

  print('Publishing text note...');
  nostr.relaysService.sendEventToRelays(textNote);
  print('Text note published.');

  print('Publishing metadata...');
  nostr.relaysService.sendEventToRelays(metadataEvent);
  print('Metadata published.');
  print('');

  // ---------------------------------------------------------------------------
  // 5. Subscribing to Events
  // ---------------------------------------------------------------------------

  // Subscribe to recent text notes
  print('Subscribing to text notes (kind 1)...');
  final textNoteSub = nostr.relaysService.startEventsSubscription(
    request: NostrRequest(
      filters: [
        NostrFilter(
          kinds: [1],
          limit: 10,
        ),
      ],
    ),
  );

  // Listen to events for a limited time
  final completer = Completer<void>();
  int eventCount = 0;

  final streamSub = textNoteSub.stream.listen((NostrEvent event) {
    eventCount++;
    print('--- Event #$eventCount ---');
    print('  Author:  ${event.pubkey?.substring(0, 16)}...');
    print('  Content: ${(event.content ?? "").length > 80 ? event.content!.substring(0, 80) + "..." : event.content}');
    print('  Created: ${DateTime.fromMillisecondsSinceEpoch((event.createdAt ?? 0) * 1000)}');

    if (eventCount >= 10) {
      completer.complete();
    }
  });

  // Wait for events or timeout after 15 seconds
  await Future.any([
    completer.future,
    Future.delayed(const Duration(seconds: 15)),
  ]);

  print('\nReceived $eventCount events.');
  print('');

  // ---------------------------------------------------------------------------
  // 6. Filtered Subscriptions
  // ---------------------------------------------------------------------------

  // Subscribe to events from a specific author
  print('Subscribing to own events...');
  final authorSub = nostr.relaysService.startEventsSubscription(
    request: NostrRequest(
      filters: [
        NostrFilter(
          authors: [keyPair.public],
          kinds: [1],
          limit: 5,
        ),
      ],
    ),
  );

  // Subscribe to events with a specific hashtag
  print('Subscribing to #nostr hashtag...');
  final hashtagSub = nostr.relaysService.startEventsSubscription(
    request: NostrRequest(
      filters: [
        NostrFilter(
          t: ['nostr'],  // NIP-12 generic tag query
          kinds: [1],
          limit: 10,
        ),
      ],
    ),
  );

  // Multiple filters in one subscription (OR logic)
  final combinedSub = nostr.relaysService.startEventsSubscription(
    request: NostrRequest(
      filters: [
        NostrFilter(kinds: [1], limit: 10),     // text notes
        NostrFilter(kinds: [7], limit: 10),     // reactions
        NostrFilter(kinds: [30023], limit: 5),  // long-form content
      ],
    ),
  );

  // Let subscriptions run briefly
  await Future.delayed(const Duration(seconds: 5));

  // ---------------------------------------------------------------------------
  // 7. NIP-05 Verification
  // ---------------------------------------------------------------------------

  print('Verifying NIP-05 identity...');
  try {
    final nip05Result = await nostr.utilsService.verifyNip05(
      internetIdentifier: 'jb55@jb55.com',
    );
    if (nip05Result != null) {
      print('NIP-05 verified: jb55@jb55.com');
    } else {
      print('NIP-05 verification failed.');
    }
  } catch (e) {
    print('NIP-05 lookup error: $e');
  }
  print('');

  // ---------------------------------------------------------------------------
  // 8. NIP-19 Encoding
  // ---------------------------------------------------------------------------

  final noteEncoded = nostr.utilsService.encodeNoteId(textNote.id!);
  print('note1...: $noteEncoded');

  // Decode NIP-19 entities
  // final decoded = nostr.utilsService.decodeNoteId('note1...');
  print('');

  // ---------------------------------------------------------------------------
  // 9. Cleanup
  // ---------------------------------------------------------------------------

  // Close individual subscriptions
  await streamSub.cancel();
  nostr.relaysService.closeEventsSubscription(textNoteSub.subscriptionId);
  nostr.relaysService.closeEventsSubscription(authorSub.subscriptionId);
  nostr.relaysService.closeEventsSubscription(hashtagSub.subscriptionId);
  nostr.relaysService.closeEventsSubscription(combinedSub.subscriptionId);

  print('All subscriptions closed. Done.');
}


// =============================================================================
// Section 2: Alternative -- Using NDK (relaystr/ndk)
// =============================================================================

/// NDK example (higher-level API with inbox/outbox support).
///
/// pubspec.yaml:
///   ndk: ^0.7.1

/*
import 'package:ndk/ndk.dart';

Future<void> ndkExample() async {
  // Initialize with a BIP-340 signer
  final ndk = Ndk(NdkConfig(
    eventSigner: Bip340EventSigner(
      privateKey: 'your_hex_private_key',
      publicKey: 'your_hex_public_key',
    ),
    cache: MemCacheManager(),
  ));

  // Fetch metadata (uses inbox/outbox automatically)
  final metadata = await ndk.metadata.loadMetadata('target_pubkey_hex');
  print('Name: ${metadata?.name}');
  print('About: ${metadata?.about}');

  // Broadcast a text note
  await ndk.broadcast.broadcastTextNote(
    content: 'Hello from NDK!',
  );

  // Query events
  final response = ndk.requests.query(
    filters: [
      Filter(kinds: [1], limit: 20),
    ],
  );

  await for (final event in response.stream) {
    print('${event.pubKey}: ${event.content}');
  }

  // Fetch and display contact list
  final contacts = await ndk.follows.getContactList('pubkey_hex');
  print('Following ${contacts.length} accounts');
}
*/


// =============================================================================
// Section 3: Alternative -- Using dart-nostr (ethicnology/dart-nostr)
// =============================================================================

/// dart-nostr example (lower-level, includes NIP-44 encryption).
///
/// pubspec.yaml:
///   nostr: ^1.4.0

/*
import 'package:nostr/nostr.dart';

void dartNostrExample() {
  // Generate keys
  final keys = Keys.generate();
  print('Public key: ${keys.public}');
  print('Private key: ${keys.secret}');

  // Create and sign an event
  final event = Event.from(
    kind: 1,
    content: 'Hello from dart-nostr!',
    privkey: keys.secret,
    tags: [
      ['t', 'nostr'],
    ],
  );

  print('Serialized event: ${event.serialize()}');

  // Create a subscription request
  final subscriptionId = generate64RandomHexChars();
  final request = Request(subscriptionId, [
    Filter(
      kinds: [1],
      limit: 20,
    ),
  ]);

  print('Serialized request: ${request.serialize()}');

  // NIP-44 encryption (modern standard)
  final recipientKeys = Keys.generate();

  final encrypted = Nip44.encrypt(
    'This is a secret message',
    keys.secret,
    recipientKeys.public,
  );
  print('Encrypted: $encrypted');

  final decrypted = Nip44.decrypt(
    encrypted,
    recipientKeys.secret,
    keys.public,
  );
  print('Decrypted: $decrypted');
}
*/


// =============================================================================
// Section 4: Flutter Widget Example (conceptual)
// =============================================================================

/// A complete Flutter screen showing a Nostr feed.
/// Requires: flutter, dart_nostr, flutter_riverpod

/*
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:dart_nostr/dart_nostr.dart';

// Provider for the event feed
final feedProvider = StreamProvider<NostrEvent>((ref) {
  final nostr = Nostr.instance;
  final subscription = nostr.relaysService.startEventsSubscription(
    request: NostrRequest(
      filters: [NostrFilter(kinds: [1], limit: 50)],
    ),
  );
  return subscription.stream;
});

// Collected events
final eventsProvider = StateNotifierProvider<EventsNotifier, List<NostrEvent>>(
  (ref) {
    final notifier = EventsNotifier();
    ref.listen(feedProvider, (prev, next) {
      next.whenData((event) => notifier.addEvent(event));
    });
    return notifier;
  },
);

class EventsNotifier extends StateNotifier<List<NostrEvent>> {
  EventsNotifier() : super([]);

  void addEvent(NostrEvent event) {
    state = [event, ...state].take(200).toList();
  }
}

class NostrFeedScreen extends ConsumerWidget {
  const NostrFeedScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final events = ref.watch(eventsProvider);

    return Scaffold(
      appBar: AppBar(title: const Text('Nostr Feed')),
      body: events.isEmpty
          ? const Center(child: CircularProgressIndicator())
          : ListView.builder(
              itemCount: events.length,
              itemBuilder: (context, index) {
                final event = events[index];
                return Card(
                  margin: const EdgeInsets.symmetric(
                    horizontal: 12,
                    vertical: 6,
                  ),
                  child: ListTile(
                    leading: CircleAvatar(
                      child: Text(
                        (event.pubkey ?? '??').substring(0, 2).toUpperCase(),
                      ),
                    ),
                    title: Text(
                      event.content ?? '',
                      maxLines: 3,
                      overflow: TextOverflow.ellipsis,
                    ),
                    subtitle: Text(
                      '${event.pubkey?.substring(0, 12)}... - '
                      '${DateTime.fromMillisecondsSinceEpoch((event.createdAt ?? 0) * 1000)}',
                    ),
                  ),
                );
              },
            ),
    );
  }
}
*/

File: nostr-knowledge/libraries/rust/examples/basic_usage.rs
// Basic Usage Example for nostr-sdk
//
// Demonstrates: key generation, event creation, connecting to relays,
// publishing events, subscribing to events, and receiving events.
//
// Cargo.toml dependencies:
//   nostr-sdk = "0.38"
//   tokio = { version = "1", features = ["full"] }
//   tracing-subscriber = "0.3"
//
// Run: cargo run --example basic_usage

use std::time::Duration;

use nostr_sdk::prelude::*;

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging (optional but helpful for debugging)
    tracing_subscriber::fmt::init();

    // ---------------------------------------------------------------
    // 1. KEY GENERATION AND MANAGEMENT
    // ---------------------------------------------------------------

    // Generate a brand new keypair
    let keys = Keys::generate();
    println!("=== Key Generation ===");
    println!("Public key (hex):    {}", keys.public_key().to_hex());
    println!("Public key (bech32): {}", keys.public_key().to_bech32()?);
    println!("Secret key (bech32): {}", keys.secret_key().to_bech32()?);
    println!();

    // You can also parse an existing secret key:
    // let keys = Keys::parse("nsec1...")?;
    // let keys = Keys::parse("hex_secret_key_here")?;

    // ---------------------------------------------------------------
    // 2. BUILD A CLIENT AND CONNECT TO RELAYS
    // ---------------------------------------------------------------

    let client = Client::builder().signer(keys.clone()).build();

    // Add relays (connection is not started yet)
    client.add_relay("wss://relay.damus.io").await?;
    client.add_relay("wss://nos.lol").await?;
    client.add_relay("wss://relay.rip").await?;

    println!("=== Connecting to Relays ===");
    client.connect().await;
    println!("Connected to {} relays", client.relays().await.len());
    println!();

    // ---------------------------------------------------------------
    // 3. SET PROFILE METADATA
    // ---------------------------------------------------------------

    let metadata = Metadata::new()
        .name("nostr-sdk-example-bot")
        .about("A basic example bot built with rust-nostr")
        .nip05("bot@example.com");

    let output = client
        .send_event_builder(EventBuilder::metadata(&metadata))
        .await?;
    println!("=== Profile Metadata ===");
    println!("Metadata event ID: {}", output.id().to_bech32()?);
    println!("Sent to {} relays", output.success.len());
    println!();

    // ---------------------------------------------------------------
    // 4. PUBLISH A TEXT NOTE
    // ---------------------------------------------------------------

    let builder = EventBuilder::text_note("Hello from rust-nostr! This is a basic usage example.")
        .tag(Tag::hashtag("nostr"))
        .tag(Tag::hashtag("rust"));

    let output = client.send_event_builder(builder).await?;
    let note_id = output.id();
    println!("=== Published Text Note ===");
    println!("Note ID: {}", note_id.to_bech32()?);
    println!("Accepted by: {:?}", output.success);
    if !output.failed.is_empty() {
        println!("Rejected by: {:?}", output.failed);
    }
    println!();

    // ---------------------------------------------------------------
    // 5. PUBLISH A TEXT NOTE WITH PROOF-OF-WORK
    // ---------------------------------------------------------------

    let builder = EventBuilder::text_note("This note has proof-of-work attached").pow(16);
    let output = client.send_event_builder(builder).await?;
    println!("=== POW Text Note ===");
    println!("POW Note ID: {}", output.id().to_bech32()?);
    println!();

    // ---------------------------------------------------------------
    // 6. REACT TO AN EVENT
    // ---------------------------------------------------------------

    // Create a reaction to our own note
    let event_to_fetch = client.sign_event_builder(
        EventBuilder::text_note("Hello from rust-nostr! This is a basic usage example.")
    ).await?;

    let builder = EventBuilder::reaction(&event_to_fetch, "+");
    client.send_event_builder(builder).await?;
    println!("=== Reaction ===");
    println!("Reacted to: {}", event_to_fetch.id.to_bech32()?);
    println!();

    // ---------------------------------------------------------------
    // 7. FETCH EVENTS (short-lived query)
    // ---------------------------------------------------------------

    println!("=== Fetching Recent Text Notes ===");

    let filter = Filter::new()
        .kind(Kind::TextNote)
        .limit(5);

    let events = client
        .fetch_events(filter)
        .timeout(Duration::from_secs(10))
        .await?;

    println!("Fetched {} events:", events.len());
    for event in events.iter().take(5) {
        let content_preview = if event.content.len() > 80 {
            format!("{}...", &event.content[..80])
        } else {
            event.content.clone()
        };
        println!(
            "  [{}] {}: {}",
            event.created_at,
            &event.pubkey.to_hex()[..8],
            content_preview,
        );
    }
    println!();

    // ---------------------------------------------------------------
    // 8. STREAM EVENTS (short-lived subscription with async iterator)
    // ---------------------------------------------------------------

    println!("=== Streaming Text Notes (10 seconds) ===");

    let filter = Filter::new()
        .kind(Kind::TextNote)
        .since(Timestamp::now());

    let mut stream = client
        .stream_events(filter)
        .timeout(Duration::from_secs(10))
        .policy(ReqExitPolicy::WaitForTimeout)
        .await?;

    let mut count = 0;
    while let Some((relay_url, result)) = stream.next().await {
        match result {
            Ok(event) => {
                count += 1;
                let content_preview = if event.content.len() > 60 {
                    format!("{}...", &event.content[..60])
                } else {
                    event.content.clone()
                };
                println!(
                    "  [{}] from {}: {}",
                    relay_url,
                    &event.pubkey.to_hex()[..8],
                    content_preview,
                );

                // Stop after 10 events for this demo
                if count >= 10 {
                    break;
                }
            }
            Err(e) => {
                eprintln!("  Error: {}", e);
            }
        }
    }
    println!("Streamed {} events", count);
    println!();

    // ---------------------------------------------------------------
    // 9. LONG-LIVED SUBSCRIPTION (via notifications)
    // ---------------------------------------------------------------

    println!("=== Long-lived Subscription (15 seconds) ===");

    let filter = Filter::new()
        .kind(Kind::TextNote)
        .since(Timestamp::now());

    let Output { val: sub_id, .. } = client.subscribe(filter).await?;
    println!("Subscription ID: {}", sub_id);

    let mut notifications = client.notifications();
    let deadline = Timestamp::now().as_u64() + 15;

    while let Some(notification) = notifications.next().await {
        // Exit after 15 seconds
        if Timestamp::now().as_u64() > deadline {
            break;
        }

        match notification {
            ClientNotification::Event {
                subscription_id,
                event,
                relay_url,
            } => {
                if subscription_id == sub_id {
                    let content_preview = if event.content.len() > 60 {
                        format!("{}...", &event.content[..60])
                    } else {
                        event.content.clone()
                    };
                    println!(
                        "  [{}] {}: {}",
                        relay_url,
                        &event.pubkey.to_hex()[..8],
                        content_preview,
                    );
                }
            }
            ClientNotification::Message { relay_url, message } => {
                if let RelayMessage::Notice(notice) = message {
                    println!("  Notice from {}: {}", relay_url, notice);
                }
            }
            _ => {}
        }
    }

    // Clean up: close subscription
    client.unsubscribe(&sub_id).await?;
    println!("Subscription closed");
    println!();

    // ---------------------------------------------------------------
    // 10. DELETE AN EVENT
    // ---------------------------------------------------------------

    let delete_request = EventDeletionRequest::new()
        .id(note_id)
        .reason("cleaning up example events");
    let builder = EventBuilder::delete(delete_request);
    client.send_event_builder(builder).await?;
    println!("=== Cleanup ===");
    println!("Deletion request sent for: {}", note_id.to_bech32()?);

    // ---------------------------------------------------------------
    // SHUTDOWN
    // ---------------------------------------------------------------

    client.shutdown().await;
    println!("Client shut down. Done!");

    Ok(())
}
